langserver/symbol.go

```diff
@@ -239,17 +250,41 @@ func score(q Query, s lsp.SymbolInformation) (scor int) {
 
 // toSym returns a SymbolInformation value derived from values we get
 // from the Go parser and doc packages.
-func toSym(name, container string, kind lsp.SymbolKind, fs *token.FileSet, pos token.Pos) lsp.SymbolInformation {
-	container = filepath.Base(container)
-	if i := strings.LastIndex(container, " "); i >= 0 {
-		container = container[i+1:]
+func toSym(name string, bpkg *build.Package, recv string, kind lsp.SymbolKind, fs *token.FileSet, pos token.Pos) symbolPair {
+	container := recv
+	if container == "" {
+		container = filepath.Base(bpkg.ImportPath)
+	}
+	return symbolPair{
+		SymbolInformation: lsp.SymbolInformation{
+			Name:          name,
+			Kind:          kind,
+			Location:      goRangeToLSPLocation(fs, pos, pos+token.Pos(len(name))-1),
+			ContainerName: container,
+		},
+		// NOTE: fields must be kept in sync with workspace_refs.go:defSymbolDescriptor
+		desc: lspext.SymbolDescriptor{
+			"vendor":      IsVendorDir(bpkg.Dir),
+			"package":     path.Clean(bpkg.ImportPath),
+			"packageName": bpkg.Name,
+			"recv":        recv,
+			"name":        name,
+			"id":          fmt.Sprintf("%s:%s:%s:%s", path.Clean(bpkg.ImportPath), bpkg.Name, recv, name),
+		},
 	}
-	return lsp.SymbolInformation{
-		Name:          name,
-		Kind:          kind,
-		Location:      goRangeToLSPLocation(fs, pos, pos+token.Pos(len(name))-1),
-		ContainerName: container,
+}
+
+// symbolContains tells if a exactly contains b.
+//
+// The only exception is the `id` field, which is treated specially. Instead of
+// a byte-by-byte string comparison, idEqual is used.
+func symbolContains(a, b lspext.SymbolDescriptor) bool {
+	for k, v := range a {
+		if bv, ok := b[k]; !ok || bv != v {
```

Comment:
@felixfbecker yeah I was actually commenting on interface equality. I thought golang would actually just compare interfaces objects (ie it would be like a pointer), but comparison operators on interfaces actually do something like
```
return a.Type == b.Type && a.ActualValue == b.ActualValue
```
TIL